### PR TITLE
Check ib point outside

### LINKED ISSUE #1390: provide clearer error messages for structure initialization errors
For IBMethod, all of the initial positions need to be inside the computational domain. We should provide a clear error message if any of the points are not in the domain during initialization.

### PR DESCRIPTION
**Pull request by drwells**
Fixes #1390. I allow points to be outside the domain as long as it occurs in a periodic direction, since those are (logically) still inside the domain.

### IBAMR Pull Request Checklist
- [x] Does the test suite pass?
- [x] Was clang-format run?
- [x] Were relevant issues cited? Please remember to add `Fixes #12345` to close
      the issue automatically if we are fixing the problem.
- [x] Is this a change others will want to know about? If so, then has a
      changelog entry been added?
- [x] If new data structures have been added to a class then have they been set
      up appropriately for restarts? If so, ensure that the restart version
      number is incremented.
- [x] Does this change include a bug fix or new functionality? If so, a new test
      or tests should be added. New tests should run quickly (less than a minute
      in release mode). If possible, an older test should gain a new option so
      that we do not need to compile more test executables.
- [x] Did you (if your account has permission to do so) set relevant labels on
      GitHub for the pull request?

### CODE REVIEW

File: src/IB/IBRedundantInitializer.cpp [hunk 1]
```diff
@@ -147,6 +147,43 @@ IBRedundantInitializer::getLevelHasLagrangianData(const int level_number, const
...
+            {
+                // Points that are inside the domain with periodicity are OK -
+                // exclude points that are truly outside.
+                std::pair<int, int> point_index(j, k);
+                const Point& X = getShiftedVertexPosn(
+                    point_index, vertex_level_number, domain_x_lower, domain_x_upper, periodic_shift);
+
+                for (int d = 0; d < NDIM; ++d)
+                {
+                    if ((X[d] < domain_x_lower[d]) || (domain_x_upper[d] < X[d]))
```
Comment (boyceg) [inline src/IB/IBRedundantInitializer.cpp:175]: Are there any tests that intentionally put IB points right on the domain boundary? It would be good to confirm that it is OK to do this.

At least away from computational domain boundaries, IB points are assigned to patches if patch_lower[d] <= X[d] < patch_upper[d]. We should confirm that this is handled correctly for patches that abut the domain boundary.

File: src/IB/IMPInitializer.cpp [hunk 2]
```diff
@@ -250,6 +250,48 @@ IMPInitializer::getLevelHasLagrangianData(const int level_number, const bool /*c
...
+            Pointer<Patch<NDIM> > patch = level->getPatch(p());
+            std::vector<std::pair<int, int> > patch_vertices; // TODO do we need this call?
+            getPatchVertices(patch_vertices, patch, ln, false);
+            for (const auto& point_idx : patch_vertices)
+            {
+                // Get the coordinates of the present vertex.
+                const libMesh::Point& X = getVertexPosn(point_idx, ln);
+                for (int d = 0; d < NDIM; ++d)
+                {
+                    if (X(d) < grid_x_lower[d])
```
Comment (boyceg) [inline src/IB/IMPInitializer.cpp:275]: Same comment as above.

File: tests/CMakeLists.txt [hunk 3]
```diff
@@ -18,7 +18,7 @@ ADD_CUSTOM_TARGET(tests)
 #    only compiles the IBFE tests.
 # 2. all input and output files in these source directories are symlinked into
 #    their corresponding build directories.
-SET(TEST_DIRECTORIES CIB ConstraintIB IB IBFE IIM IBTK adv_diff advect coarsen
+SET(TEST_DIRECTORIES CIB ConstraintIB IB IBFE IIM IMP IBTK adv_diff advect coarsen
```
Comment (boyceg) [inline tests/CMakeLists.txt:21]: 👍

**Review by boyceg — COMMENTED** [review]
This looks good. We should confirm that it is OK to put points right on the lower/upper boundaries of the computational domain, or require them always to be inside the computational domain.

(`LDataManager` is set up so that it does "something reasonable" if points try to _leave_ the domain. This would be outside the scope of this PR, but it might also be good to add a test for that case ---passively advected particles that are very close to the domain boundary and that should be pushed out of the domain in one or two time steps.)

**Comment by drwells** [comment]
I checked and things work with both points on periodic and physical boundaries. Here's a picture:

visit0000


I also added some tests showing that this works correctly in serial and parallel.

**Comment by drwells** [comment]
thats pretty easy - I can just move my quarter box to the top right.

**Review by boyceg — APPROVED** [review]
Looks good. Feel free to merge as soon as all corner cases are covered in the tests.

**Comment by drwells** [comment]
I'll get that done once infrastructure week is complete :)

File: src/IB/IBRedundantInitializer.cpp [hunk 4]
```diff
@@ -1370,8 +1414,26 @@ IBRedundantInitializer::getPatchVerticesAtLevel(std::vector<std::pair<int, int>
...
+                    if (X[d] == domain_x_upper[d])
+                    {
+                        idx(d) -= 1;
+                    }
+                }
+            }
+            if (patch_box.contains(idx))
+            {
+                patch_vertices.push_back(point_index);
+            }
```
Comment (drwells) [inline src/IB/IBRedundantInitializer.cpp:1436]: We now permit points that are on the upper part of the domain - here's the logic for that case. Two of the new tests verify that this works.

**Comment by drwells** [comment]
@boyceg since I force-pushed our previous discussion isn't visible. Are you generally happy with the choices here? If so lets merge.

**Comment by boyceg** [comment]
Looks good to me.
